src/confirmation.rs

```diff
@@ -0,0 +1,79 @@
+use crate::zulip::api::ZulipApi;
+use crate::{get_env, run_diffs, ServiceDiff};
+use sha2::{Digest, Sha256};
+
+pub(crate) struct Confirmation {
+    zulip: ZulipApi,
+    stream: String,
+    topic: String,
+    base_url: String,
+
+    diffs: Vec<ServiceDiff>,
+    hash: String,
+}
+
+impl Confirmation {
+    pub(crate) fn new(diffs: Vec<ServiceDiff>) -> anyhow::Result<Self> {
+        let mut hash = Sha256::new();
+        hash.update(&serde_json::to_vec(&diffs).unwrap());
+        let hash = hex::encode(hash.finalize());
+
+        Ok(Self {
+            zulip: ZulipApi::new(
+                get_env("ZULIP_USERNAME")?,
+                get_env("ZULIP_API_TOKEN")?,
+                false,
+            ),
+            stream: get_env("CONFIRMATION_STREAM")?,
+            topic: get_env("CONFIRMATION_TOPIC")?,
+            base_url: get_env("CONFIRMATION_BASE_URL")?,
+            diffs,
+            hash,
+        })
+    }
+
+    pub(crate) fn run(self) -> anyhow::Result<()> {
+        if let Ok(expected) = get_env("CONFIRMATION_EXPECTED_HASH") {
```

Comment:
Nit: I find it confusing that this method is opaquely deciding when to run the diffs and when to post for approval. I would find it more understandable if the logic for that were spelled out more clearly than "is a env var present?".